esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5892,260 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+             */
+            break;
+        }
+        case CFileAccessRole_External:
+        {
+            IArrayOf<IEspDFUPartLocations> dfuPartLocations;
+            IArrayOf<IEspDFUPartCopies> dfuPartCopies;
+            resp.setNumParts(getFilePartsInfo(context, df, req.getCluster(), dfuPartLocations, dfuPartCopies));
+            resp.setFilePartLocations(dfuPartLocations);
+            resp.setFileParts(dfuPartCopies);
+            if (req.getReturnJsonTypeInfo() || req.getReturnJsonTypeInfo())
+            {
+                MemoryBuffer binLayout;
+                StringBuffer jsonLayout;
+                if (!getRecordFormatFromRtlType(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo()))
+                    getRecordFormatFromECL(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo());
+                if (req.getReturnJsonTypeInfo() && jsonLayout.length())
+                    resp.setRecordTypeInfoJson(jsonLayout.str());
+                if (req.getReturnBinTypeInfo() && binLayout.length())
+                    resp.setRecordTypeInfoBin(binLayout);
+            }
+            break;
+        }
+        default:
+            throwUnexpected();
+    }
+
+    // setup "expiryTime"
+    unsigned expirySecs = req.getExpirySeconds();
+    if (expirySecs > fileAccessExpirySeconds)
+        expirySecs = fileAccessExpirySeconds;
+    time_t now;
+    time(&now);
+    CDateTime expiryDt;
+    expiryDt.set(now + expirySecs);
+    StringBuffer expiryTimeStr;
+    expiryDt.getString(expiryTimeStr);
+
+    StringBuffer keyPairName;
+    getFileDafilesrvKeyName(keyPairName, *df);
+
+    Owned<IPropertyTree> metaInfo = createPTree();
+    getFileMeta(*metaInfo, *df, udesc, role, expiryTimeStr, keyPairName, req);
+    StringBuffer metaInfoJsonBlob;
+    toJSON(metaInfo, metaInfoJsonBlob);
+    resp.setMetaInfoBlob(metaInfoJsonBlob);
+    resp.setKeyName(keyPairName);
+    resp.setExpiryTime(expiryTimeStr);
+}
+
+
+SecAccessFlags translateToSecAccessFlags(CSecAccessType from)
+{
+    switch (from)
+    {
+        case CSecAccessType_Access:
+            return SecAccess_Access;
+        case CSecAccessType_Read:
+            return SecAccess_Read;
+        case CSecAccessType_Write:
+            return SecAccess_Write;
+        case CSecAccessType_Full:
+            return SecAccess_Full;
+        case CSecAccessType_None:
+        default:
+            return SecAccess_None;
+    }
+}
+
+bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &req, IEspDFUFileAccessResponse &resp)
+{
+    try
+    {
+        SecAccessFlags accessType = translateToSecAccessFlags(req.getAccessType());
+        if (SecAccess_None == accessType)
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "onDFUFileAccess - Permission denied.");
+        else
```

Comment:
No else needed here

Reply:
strictly no, but this is a style thing only, arguably it is clearer that the else is there, in fact I prefer it changed to:
```
        if (SecAccess_None == accessType)
            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "onDFUFileAccess - Permission denied.");
        else if (!context.validateFeatureAccess(FEATURE_URL, accessType, false))
            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "onDFUFileAccess - Permission denied.");
```